edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java

```diff
@@ -120,44 +120,52 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
             return false;
         }
 
-
         var bpn = participantAgent.getIdentity();
         var groups = store.resolveForBpn(bpn);
 
         // BPN not found in database
         if (groups.failed()) {
-            policyContext.reportProblem(groups.getFailureDetail());
-            return false;
+            return evaluateEmptyGroupsResponse(operator, policyContext, groups.getFailureDetail());
         }
+
         var assignedGroups = groups.getContent();
 
         // BPN was found, but it does not have groups assigned.
         if (assignedGroups.isEmpty()) {
-            policyContext.reportProblem("No groups were assigned to BPN " + bpn);
-            return false;
+            return evaluateEmptyGroupsResponse(operator, policyContext, "No groups were assigned to BPN " + bpn);
         }
 
         // right-operand is anything other than String or Collection
         var rightOperand = parseRightOperand(rightValue, policyContext);
-        if (rightOperand == null) {
+        if (rightOperand.isEmpty()) {
             return false;
         }
 
         //call evaluator function
         return OPERATOR_EVALUATOR_MAP.get(operator).apply(new BpnGroupHolder(assignedGroups, rightOperand));
     }
 
+    private boolean evaluateEmptyGroupsResponse(final Operator operator, final PolicyContext policyContext,
```

Comment:
this method is quite obscure, from reading the signature it is not clear which combinations of parameters produce which result. Its fine to have a method to encapsulate error handling, but this is not whats happening here.
The method name is misleading as well. Please rewrite this, and a word of documentation can't hurt either.

Reply:
I refactored the method and added a documentation to be more clear